src/sinks/prometheus/exporter.rs

```diff
@@ -302,175 +434,93 @@ impl PrometheusExporter {
 impl StreamSink for PrometheusExporter {
     async fn run(mut self: Box<Self>, mut input: BoxStream<'_, Event>) -> Result<(), ()> {
         self.start_server_if_needed().await;
+
+        let mut last_flush = Instant::now();
+        let mut normalizer = MetricNormalizer::from_inner(PrometheusExporterMetricNormalizer {
+            distributions_as_summaries: self.config.distributions_as_summaries,
+            buckets: self.config.buckets.clone(),
+        });
+
         while let Some(event) = input.next().await {
-            let item = event.into_metric();
-            let mut metrics = self.metrics.write().unwrap();
-
-            // sets need to be expired from time to time
-            // because otherwise they could grow infinitely
-            let now = Utc::now().timestamp();
-            let interval = now - metrics.last_flush_timestamp;
-            if interval > self.config.flush_period_secs as i64 {
-                metrics.last_flush_timestamp = now;
-
-                let now = Instant::now();
-                metrics.map = metrics
-                    .map
-                    .drain(..)
-                    .map(|(mut entry, metadata)| {
-                        if metadata.is_incremental_set {
-                            entry.0.zero();
-                        }
-                        (entry, metadata)
-                    })
-                    .filter(|(_metric, metadata)| {
-                        now.duration_since(metadata.updated_at).as_secs()
-                            < self.config.flush_period_secs
-                    })
-                    .collect();
+            let metric = event.into_metric();
+
+            if let Some(normalized) = normalizer.apply(metric) {
+                // Push this metric into the expiration tracker.  All metrics that we handle here
+                // have already been normalized into absolute metrics, so we're really just updating
+                // the "last updated" field for existing/tracked metrics, and add it in for
+                // not-yet-seen ones.
+                let mut metrics = self.metrics.write().unwrap();
+
+                let metric_ref = MetricRef::from_metric(&normalized);
+                match metrics.get_mut(&metric_ref) {
+                    Some((data, metadata)) => {
+                        *data = normalized;
+                        metadata.last_updated = Instant::now();
+                    }
+                    None => {
+                        // We specifically clear sets when they expire rather than totally delete
+                        // them, and this flag is how we know to do that.
+                        let should_clear = matches!(normalized.value(), MetricValue::Set { .. });
+                        metrics.insert(metric_ref, (normalized, MetricMetadata::new(should_clear)));
+                    }
+                }
```

Comment:
Can this be simplified with the `entry` interface? ie `metrics.entry(metric_ref).and_modify(…).or_insert_with(…)`

Reply:
I tried that at first, but it involves an extra clone because you have to move the metric into both the `and_modify` closure and the `or_insert_with` closure.

I swore there was a way to avoid this, but I guess it's really only doable for the key via `.entry(..)`.

Reply:
Yeah, I think I was looking more at the key and less at the rest.